File: 1_GuidedExercise/idea_09.rb
class Idea09
  # Can you use inline and ide hints make this code simpler?
  def overly_obfuscated(n)
    first_number = 1
    second_number = 1

    if is_first_number(n)
      return first_number
    elsif is_second_number(n)
      return second_number
    else
      return get_general_fibonacci_number(n)
    end
  end

  def get_general_fibonacci_number(n)
    get_previous_fibonacci_number(n) + get_previous_previous_fibonacci_number(n)
  end

  def get_previous_previous_fibonacci_number(n)
    get_mth_previous_fibonacci_number(n, 2)
  end

  def get_previous_fibonacci_number(n)
    get_mth_previous_fibonacci_number(n, 1)
  end

  def get_mth_previous_fibonacci_number(n, m)
    self.overly_obfuscated(n - m)
  end

  def is_second_number(n)
    n == 2
  end

  def is_first_number(n)
    n == 1
  end
end

File: js/1_GuidedExercise/src/idea_09.js
// Task: Use inline variable, inline method and ide hints make this code simpler.
const overly_obfuscated = (n) => {
  const firstNumber = 1
  const secondNumber = 1

  if (isFirstNumber(n)) {
    return firstNumber
  } else if (isSecondNumber(n)) {
    return secondNumber
  } else {
    return getGeneralFibonacciNumber(n)
  }
}

function getGeneralFibonacciNumber(n) {
  return getPreviousFibonacciNumber(n) + getPreviousPreviousFibonacciNumber(n);
}

function getPreviousPreviousFibonacciNumber(n) {
  return getMthPreviousFibonacciNumber(n, 2);
}

function getPreviousFibonacciNumber(n) {
  return getMthPreviousFibonacciNumber(n, 1);
}

function getMthPreviousFibonacciNumber(n, m) {
  return overly_obfuscated(n-m)
}

function isSecondNumber(n) {
  return n === 2
}

function isFirstNumber(n) {
  return n === 1
}

module.exports = { overly_obfuscated }
